File: goldenmatch/core/golden.py
"""Golden record builder with per-field merge strategies."""

from __future__ import annotations

from collections import Counter
from dataclasses import dataclass, field as dataclass_field
from typing import Any

import polars as pl

from goldenmatch.config.schemas import GoldenFieldRule, GoldenRulesConfig


@dataclass
class FieldProvenance:
    value: Any
    source_row_id: int
    strategy: str
    confidence: float
    candidates: list[dict] = dataclass_field(default_factory=list)


@dataclass
class ClusterProvenance:
    cluster_id: int
    cluster_quality: str
    cluster_confidence: float
    fields: dict[str, FieldProvenance] = dataclass_field(default_factory=dict)


@dataclass
class GoldenRecordResult:
    df: pl.DataFrame
    provenance: list[ClusterProvenance] = dataclass_field(default_factory=list)

# Columns to skip when building golden records
_INTERNAL_PREFIXES = ("__row_id__", "__source__", "__block_key__", "__mk_")


def _is_internal(col: str) -> bool:
    return any(col.startswith(p) for p in _INTERNAL_PREFIXES) or col == "__mk_"


def merge_field(
    values: list,
    rule: GoldenFieldRule,
    sources: list[str] | None = None,
    dates: list | None = None,
    quality_weights: list[float] | None = None,
) -> tuple[object, float, int | None]:
    """Merge a list of values using the given rule's strategy.

    Returns (value, confidence, source_index) where source_index is the
    index into the values list that the winning value came from.
    """
    non_null = [(i, v) for i, v in enumerate(values) if v is not None]

    if not non_null:
        return (None, 0.0, None)

    # If all non-null values are identical, return with confidence 1.0
    unique_vals = set(v for _, v in non_null)
    if len(unique_vals) == 1:
        return (non_null[0][1], 1.0, non_null[0][0])

    strategy = rule.strategy

    if strategy == "most_complete":
        return _most_complete(non_null, quality_weights)
    elif strategy == "majority_vote":
        return _majority_vote(non_null, quality_weights)
    elif strategy == "source_priority":
        return _source_priority(values, rule, sources)
    elif strategy == "most_recent":
        return _most_recent(values, dates)
    elif strategy == "first_non_null":
        return _first_non_null(non_null, quality_weights)
    else:
        raise ValueError(f"Unknown strategy: {strategy}")


def _most_complete(non_null: list[tuple[int, object]], quality_weights: list[float] | None = None) -> tuple:
    str_vals = [(i, str(v), v) for i, v in non_null]
    max_len = max(len(s) for _, s, _ in str_vals)
    longest = [(i, s, v) for i, s, v in str_vals if len(s) == max_len]
    if len(longest) == 1:
        return (longest[0][2], 1.0, longest[0][0])
    # Tie-break by quality weight if available
    if quality_weights is not None:
        best = max(longest, key=lambda x: quality_weights[x[0]] if x[0] < len(quality_weights) else 1.0)
        conf = min(1.0, 0.7 * quality_weights[best[0]]) if best[0] < len(quality_weights) else 0.7
        return (best[2], conf, best[0])
    return (longest[0][2], 0.7, longest[0][0])


def _majority_vote(non_null: list[tuple[int, object]], quality_weights: list[float] | None = None) -> tuple:
    if quality_weights is not None:
        # Weighted vote: sum quality weights per value
        value_weights: dict[object, float] = {}
        value_idx: dict[object, int] = {}
        for i, v in non_null:
            w = quality_weights[i] if i < len(quality_weights) else 1.0
            value_weights[v] = value_weights.get(v, 0.0) + w
            if v not in value_idx:
                value_idx[v] = i
        winner = max(value_weights, key=value_weights.__getitem__)
        total_weight = sum(value_weights.values())
        conf = value_weights[winner] / total_weight if total_weight > 0 else 0.0
        return (winner, conf, value_idx[winner])
    counts = Counter(v for _, v in non_null)
    winner, count = counts.most_common(1)[0]
    total = len(non_null)
    # Find the index of the first occurrence of the winner
    winner_idx = next(i for i, v in non_null if v == winner)
    return (winner, count / total, winner_idx)


def _source_priority(
    values: list,
    rule: GoldenFieldRule,
    sources: list[str] | None,
) -> tuple:
    if sources is None:
        raise ValueError("source_priority strategy requires sources list")
    source_val = {}
    source_idx = {}
    for i, (src, val) in enumerate(zip(sources, values)):
        if src not in source_val:
            source_val[src] = val
            source_idx[src] = i

    for idx, src in enumerate(rule.source_priority):
        val = source_val.get(src)
        if val is not None:
            conf = max(0.1, 1.0 - idx * 0.1)
            return (val, conf, source_idx[src])

    # Fallback: no match found in priority list
    return (None, 0.0, None)


def _most_recent(values: list, dates: list | None) -> tuple:
    if dates is None:
        raise ValueError("most_recent strategy requires dates list")
    indexed_pairs = [(i, d, v) for i, (d, v) in enumerate(zip(dates, values)) if v is not None and d is not None]
    if not indexed_pairs:
        return (None, 0.0, None)
    indexed_pairs.sort(key=lambda x: x[1], reverse=True)
    top_date = indexed_pairs[0][1]
    tied = [p for p in indexed_pairs if p[1] == top_date]
    conf = 1.0 if len(tied) == 1 else 0.5
    return (indexed_pairs[0][2], conf, indexed_pairs[0][0])


def _first_non_null(non_null: list[tuple[int, object]], quality_weights: list[float] | None = None) -> tuple:
    if quality_weights is not None:
        # Pick the non-null value with the highest quality weight
        best = max(non_null, key=lambda x: quality_weights[x[0]] if x[0] < len(quality_weights) else 1.0)
        return (best[1], 0.6, best[0])
    return (non_null[0][1], 0.6, non_null[0][0])


def build_golden_record(
    cluster_df: pl.DataFrame,
    rules: GoldenRulesConfig,
    quality_scores: dict[tuple[int, str], float] | None = None,
) -> dict:
    """Build a golden record from a cluster DataFrame.

    Returns dict of {col: {"value": v, "confidence": c}, ...,
    "__golden_confidence__": mean_of_confidences}.
    """
    result = {}
    confidences = []
    row_ids = cluster_df["__row_id__"].to_list() if "__row_id__" in cluster_df.columns else None

    for col in cluster_df.columns:
        if _is_internal(col):
            continue

        values = cluster_df[col].to_list()

        # Look up field rule or build default
        if col in rules.field_rules:
            field_rule = rules.field_rules[col]
        else:
            field_rule = GoldenFieldRule(strategy=rules.default_strategy)

        # Gather optional lists
        sources = None
        dates = None
        weights = None
        if field_rule.strategy == "source_priority" and "__source__" in cluster_df.columns:
            sources = cluster_df["__source__"].to_list()
        if field_rule.strategy == "most_recent" and field_rule.date_column:
            if field_rule.date_column in cluster_df.columns:
                dates = cluster_df[field_rule.date_column].to_list()
        if quality_scores is not None and row_ids is not None:
            weights = [quality_scores.get((rid, col), 1.0) for rid in row_ids]

        val, conf, _idx = merge_field(values, field_rule, sources=sources, dates=dates, quality_weights=weights)
        result[col] = {"value": val, "confidence": conf}
        confidences.append(conf)

    if confidences:
        result["__golden_confidence__"] = sum(confidences) / len(confidences)
    else:
        result["__golden_confidence__"] = 0.0

    return result


def build_golden_record_with_provenance(
    df: pl.DataFrame,
    rules: GoldenRulesConfig,
    clusters: dict[int, dict],
    quality_scores: dict[tuple[int, str], float] | None = None,
) -> GoldenRecordResult:
    """Build golden records with field-level provenance tracking."""
    golden_rows = []
    provenance_list = []

    cluster_col = "__cluster_id__"
    if cluster_col not in df.columns:
        # Single cluster case
        cluster_ids = [1]
        cluster_dfs = {1: df}
    else:
        cluster_ids = sorted(df[cluster_col].unique().to_list())
        cluster_dfs = {cid: df.filter(pl.col(cluster_col) == cid) for cid in cluster_ids}

    for cid in cluster_ids:
        cluster_df = cluster_dfs[cid]
        cinfo = clusters.get(cid, {})
        row_ids = cluster_df["__row_id__"].to_list() if "__row_id__" in cluster_df.columns else list(range(len(cluster_df)))

        # Build golden record + provenance in a single pass (no double merge_field call)
        field_provenance = {}
        golden_row = {"__cluster_id__": cid}
        confidences = []

        for col in cluster_df.columns:
            if _is_internal(col):
                continue
            values = cluster_df[col].to_list()
            if col in rules.field_rules:
                field_rule = rules.field_rules[col]
            else:
                field_rule = GoldenFieldRule(strategy=rules.default_strategy)

            sources = None
            dates = None
            weights = None
            if field_rule.strategy == "source_priority" and "__source__" in cluster_df.columns:
                sources = cluster_df["__source__"].to_list()
            if field_rule.strategy == "most_recent" and field_rule.date_column:
                if field_rule.date_column in cluster_df.columns:
                    dates = cluster_df[field_rule.date_column].to_list()
            if quality_scores is not None and row_ids:
                weights = [quality_scores.get((rid, col), 1.0) for rid in row_ids]

            val, conf, src_idx = merge_field(values, field_rule, sources=sources, dates=dates, quality_weights=weights)
            confidences.append(conf)

            source_row_id = row_ids[src_idx] if src_idx is not None and src_idx < len(row_ids) else row_ids[0]

            candidates = []
            for rid, v in zip(row_ids, values):
                q = quality_scores.get((rid, col), 1.0) if quality_scores else 1.0
                candidates.append({"row_id": rid, "value": v, "quality": q})

            field_provenance[col] = FieldProvenance(
                value=val,
                source_row_id=source_row_id,
                strategy=field_rule.strategy,
                confidence=conf,
                candidates=candidates,
            )
            golden_row[col] = val

        golden_row["__golden_confidence__"] = sum(confidences) / len(confidences) if confidences else 0.0
        golden_rows.append(golden_row)

        provenance_list.append(ClusterProvenance(
            cluster_id=cid,
            cluster_quality=cinfo.get("cluster_quality", "strong"),
            cluster_confidence=cinfo.get("confidence", 0.0),
            fields=field_provenance,
        ))

    golden_df = pl.DataFrame(golden_rows) if golden_rows else pl.DataFrame()
    return GoldenRecordResult(df=golden_df, provenance=provenance_list)

File: packages/goldenmatch-js/src/core/golden.ts
/**
 * golden.ts — Golden record builder with per-field merge strategies.
 * Edge-safe: no Node.js imports, pure TypeScript only.
 */

import type {
  ClusterInfo,
  ClusterProvenance,
  FieldProvenance,
  GoldenFieldRule,
  GoldenRulesConfig,
  Row,
} from "./types.js";

// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------

const INTERNAL_PREFIXES = [
  "__row_id__",
  "__source__",
  "__block_key__",
  "__mk_",
  "__cluster_id__",
  "__golden_confidence__",
];

function isInternal(col: string): boolean {
  return (
    col === "__mk_" ||
    INTERNAL_PREFIXES.some((prefix) => col.startsWith(prefix))
  );
}

// ---------------------------------------------------------------------------
// MergeField result
// ---------------------------------------------------------------------------

export interface MergeFieldResult {
  readonly value: unknown;
  readonly confidence: number;
  readonly sourceIndex: number | null;
}

// ---------------------------------------------------------------------------
// MergeField options
// ---------------------------------------------------------------------------

export interface MergeFieldOptions {
  readonly sources?: readonly string[];
  readonly dates?: readonly unknown[];
  readonly qualityWeights?: readonly number[];
}

// ---------------------------------------------------------------------------
// mergeField
// ---------------------------------------------------------------------------

/**
 * Merge a list of values using the given strategy.
 *
 * Strategies:
 * - most_complete: pick longest string value; tie-break by quality weight
 * - majority_vote: pick most frequent value; weighted by quality if available
 * - source_priority: pick first non-null from priority list
 * - most_recent: pick value with most recent date
 * - first_non_null: pick first non-null; prefer highest quality weight
 */
export function mergeField(
  values: readonly unknown[],
  rule: GoldenFieldRule,
  options?: MergeFieldOptions,
): MergeFieldResult {
  const nonNull: [number, unknown][] = [];
  for (let i = 0; i < values.length; i++) {
    if (values[i] != null) {
      nonNull.push([i, values[i]]);
    }
  }

  if (nonNull.length === 0) {
    return { value: null, confidence: 0.0, sourceIndex: null };
  }

  // All non-null values identical -> confidence 1.0 shortcut
  const uniqueVals = new Set(nonNull.map(([, v]) => v));
  if (uniqueVals.size === 1) {
    return { value: nonNull[0]![1], confidence: 1.0, sourceIndex: nonNull[0]![0] };
  }

  const strategy = rule.strategy;

  switch (strategy) {
    case "most_complete":
      return _mostComplete(nonNull, options?.qualityWeights);
    case "majority_vote":
      return _majorityVote(nonNull, options?.qualityWeights);
    case "source_priority":
      return _sourcePriority(values, rule, options?.sources);
    case "most_recent":
      return _mostRecent(values, options?.dates);
    case "first_non_null":
      return _firstNonNull(nonNull, options?.qualityWeights);
    default:
      throw new Error(`Unknown strategy: ${strategy}`);
  }
}

// ---------------------------------------------------------------------------
// Strategy implementations
// ---------------------------------------------------------------------------

function _mostComplete(
  nonNull: [number, unknown][],
  qualityWeights?: readonly number[],
): MergeFieldResult {
  const strVals = nonNull.map(
    ([i, v]) => [i, String(v), v] as [number, string, unknown],
  );
  const maxLen = Math.max(...strVals.map(([, s]) => s.length));
  const longest = strVals.filter(([, s]) => s.length === maxLen);

  if (longest.length === 1) {
    return { value: longest[0]![2], confidence: 1.0, sourceIndex: longest[0]![0] };
  }

  // Tie-break by quality weight
  if (qualityWeights) {
    const best = longest.reduce((a, b) => {
      const wa =
        a[0] < qualityWeights.length ? qualityWeights[a[0]]! : 1.0;
      const wb =
        b[0] < qualityWeights.length ? qualityWeights[b[0]]! : 1.0;
      return wa >= wb ? a : b;
    });
    const w =
      best[0] < qualityWeights.length ? qualityWeights[best[0]]! : 1.0;
    const conf = Math.min(1.0, 0.7 * w);
    return { value: best[2], confidence: conf, sourceIndex: best[0] };
  }

  return { value: longest[0]![2], confidence: 0.7, sourceIndex: longest[0]![0] };
}

function _majorityVote(
  nonNull: [number, unknown][],
  qualityWeights?: readonly number[],
): MergeFieldResult {
  if (qualityWeights) {
    // Weighted vote: sum quality weights per value
    const valueWeights = new Map<unknown, number>();
    const valueIdx = new Map<unknown, number>();

    for (const [i, v] of nonNull) {
      const w = i < qualityWeights.length ? qualityWeights[i]! : 1.0;
      valueWeights.set(v, (valueWeights.get(v) ?? 0) + w);
      if (!valueIdx.has(v)) valueIdx.set(v, i);
    }

    let winner: unknown = null;
    let bestWeight = -Infinity;
    for (const [v, w] of valueWeights) {
      if (w > bestWeight) {
        bestWeight = w;
        winner = v;
      }
    }

    let totalWeight = 0;
    for (const w of valueWeights.values()) totalWeight += w;
    const conf = totalWeight > 0 ? bestWeight / totalWeight : 0.0;
    return { value: winner, confidence: conf, sourceIndex: valueIdx.get(winner)! };
  }

  // Unweighted: count occurrences
  const counts = new Map<unknown, number>();
  for (const [, v] of nonNull) {
    counts.set(v, (counts.get(v) ?? 0) + 1);
  }

  let winner: unknown = null;
  let bestCount = -1;
  for (const [v, c] of counts) {
    if (c > bestCount) {
      bestCount = c;
      winner = v;
    }
  }

  const winnerIdx = nonNull.find(([, v]) => v === winner)![0];
  return {
    value: winner,
    confidence: bestCount / nonNull.length,
    sourceIndex: winnerIdx,
  };
}

function _sourcePriority(
  values: readonly unknown[],
  rule: GoldenFieldRule,
  sources?: readonly string[],
): MergeFieldResult {
  if (!sources) {
    throw new Error("source_priority strategy requires sources list");
  }

  const sourceVal = new Map<string, unknown>();
  const sourceIdx = new Map<string, number>();

  for (let i = 0; i < sources.length; i++) {
    const src = sources[i]!;
    if (!sourceVal.has(src)) {
      sourceVal.set(src, values[i]);
      sourceIdx.set(src, i);
    }
  }

  const priority = rule.sourcePriority ?? [];
  for (let idx = 0; idx < priority.length; idx++) {
    const src = priority[idx]!;
    const val = sourceVal.get(src);
    if (val != null) {
      const conf = Math.max(0.1, 1.0 - idx * 0.1);
      return { value: val, confidence: conf, sourceIndex: sourceIdx.get(src)! };
    }
  }

  // Fallback: no match in priority list
  return { value: null, confidence: 0.0, sourceIndex: null };
}

function _mostRecent(
  values: readonly unknown[],
  dates?: readonly unknown[],
): MergeFieldResult {
  if (!dates) {
    throw new Error("most_recent strategy requires dates list");
  }

  const indexed: [number, unknown, unknown][] = [];
  for (let i = 0; i < values.length; i++) {
    if (values[i] != null && dates[i] != null) {
      indexed.push([i, dates[i], values[i]]);
    }
  }

  if (indexed.length === 0) {
    return { value: null, confidence: 0.0, sourceIndex: null };
  }

  // Sort by date descending (works for ISO strings and numbers)
  indexed.sort((a, b) => {
    if (a[1]! > b[1]!) return -1;
    if (a[1]! < b[1]!) return 1;
    return 0;
  });

  const topDate = indexed[0]![1];
  const tied = indexed.filter(([, d]) => d === topDate);
  const conf = tied.length === 1 ? 1.0 : 0.5;

  return { value: indexed[0]![2], confidence: conf, sourceIndex: indexed[0]![0] };
}

function _firstNonNull(
  nonNull: [number, unknown][],
  qualityWeights?: readonly number[],
): MergeFieldResult {
  if (qualityWeights) {
    // Pick the non-null value with the highest quality weight
    let bestIdx = nonNull[0]![0];
    let bestVal = nonNull[0]![1];
    let bestWeight =
      nonNull[0]![0] < qualityWeights.length
        ? qualityWeights[nonNull[0]![0]]!
        : 1.0;

    for (let i = 1; i < nonNull.length; i++) {
      const [idx, val] = nonNull[i]!;
      const w = idx < qualityWeights.length ? qualityWeights[idx]! : 1.0;
      if (w > bestWeight) {
        bestWeight = w;
        bestIdx = idx;
        bestVal = val;
      }
    }

    return { value: bestVal, confidence: 0.6, sourceIndex: bestIdx };
  }

  return { value: nonNull[0]![1], confidence: 0.6, sourceIndex: nonNull[0]![0] };
}

// ---------------------------------------------------------------------------
// GoldenRecord
// ---------------------------------------------------------------------------

export interface GoldenRecord {
  readonly fields: Readonly<Record<string, { value: unknown; confidence: number }>>;
  readonly goldenConfidence: number;
}

// ---------------------------------------------------------------------------
// buildGoldenRecord
// ---------------------------------------------------------------------------

/**
 * Build a golden record from cluster rows.
 *
 * @param clusterRows - Array of row objects belonging to one cluster.
 * @param rules - Golden rules config with default strategy and field rules.
 * @param qualityScores - Optional map of `"rowId:column"` -> quality score.
 */
export function buildGoldenRecord(
  clusterRows: readonly Row[],
  rules: GoldenRulesConfig,
  qualityScores?: ReadonlyMap<string, number>,
): GoldenRecord {
  if (clusterRows.length === 0) {
    return { fields: {}, goldenConfidence: 0.0 };
  }

  // Collect all column names
  const columns = new Set<string>();
  for (const row of clusterRows) {
    for (const col of Object.keys(row)) {
      columns.add(col);
    }
  }

  const rowIds: number[] = clusterRows.map(
    (r) => (r.__row_id__ as number) ?? 0,
  );

  const fields: Record<string, { value: unknown; confidence: number }> = {};
  const confidences: number[] = [];

  for (const col of columns) {
    if (isInternal(col)) continue;

    const values = clusterRows.map((r) => r[col] ?? null);

    // Look up field rule or use default
    const fieldRule: GoldenFieldRule =
      rules.fieldRules[col] ?? { strategy: rules.defaultStrategy as GoldenFieldRule["strategy"] };

    // Gather optional lists
    let sources: string[] | undefined;
    let dates: unknown[] | undefined;
    let weights: number[] | undefined;

    if (fieldRule.strategy === "source_priority") {
      sources = clusterRows.map((r) => String(r.__source__ ?? ""));
    }
    if (fieldRule.strategy === "most_recent" && fieldRule.dateColumn) {
      dates = clusterRows.map((r) => r[fieldRule.dateColumn!] ?? null);
    }
    if (qualityScores) {
      weights = rowIds.map((rid) => qualityScores.get(`${rid}:${col}`) ?? 1.0);
    }

    const mergeOpts: MergeFieldOptions = {
      ...(sources !== undefined && { sources }),
      ...(dates !== undefined && { dates }),
      ...(weights !== undefined && { qualityWeights: weights }),
    };
    const result = mergeField(values, fieldRule, mergeOpts);
    fields[col] = { value: result.value, confidence: result.confidence };
    confidences.push(result.confidence);
  }

  const goldenConfidence =
    confidences.length > 0
      ? confidences.reduce((a, b) => a + b, 0) / confidences.length
      : 0.0;

  return { fields, goldenConfidence };
}

// ---------------------------------------------------------------------------
// buildGoldenRecordWithProvenance
// ---------------------------------------------------------------------------

export interface GoldenRecordWithProvenanceResult {
  readonly goldenRecords: readonly (Row & {
    __cluster_id__: number;
    __golden_confidence__: number;
  })[];
  readonly provenance: readonly ClusterProvenance[];
}

/**
 * Build golden records with full field-level provenance tracking.
 *
 * @param allRows - All rows with `__cluster_id__` and `__row_id__` columns.
 * @param rules - Golden rules config.
 * @param clusters - Cluster map from buildClusters.
 * @param qualityScores - Optional `"rowId:column"` -> quality score map.
 */
export function buildGoldenRecordWithProvenance(
  allRows: readonly Row[],
  rules: GoldenRulesConfig,
  clusters: ReadonlyMap<number, ClusterInfo>,
  qualityScores?: ReadonlyMap<string, number>,
): GoldenRecordWithProvenanceResult {
  // Group rows by cluster ID
  const clusterDfs = new Map<number, Row[]>();
  for (const row of allRows) {
    const cid = (row.__cluster_id__ as number) ?? 1;
    let arr = clusterDfs.get(cid);
    if (!arr) {
      arr = [];
      clusterDfs.set(cid, arr);
    }
    arr.push(row);
  }

  const clusterIds = [...clusterDfs.keys()].sort((a, b) => a - b);
  const goldenRecords: (Row & {
    __cluster_id__: number;
    __golden_confidence__: number;
  })[] = [];
  const provenanceList: ClusterProvenance[] = [];

  for (const cid of clusterIds) {
    const clusterRows = clusterDfs.get(cid)!;
    const cinfo = clusters.get(cid);
    const rowIds = clusterRows.map(
      (r) => (r.__row_id__ as number) ?? 0,
    );

    // Collect columns
    const columns = new Set<string>();
    for (const row of clusterRows) {
      for (const col of Object.keys(row)) {
        columns.add(col);
      }
    }

    const fieldProvenance: Record<string, FieldProvenance> = {};
    const goldenRow: Record<string, unknown> = { __cluster_id__: cid };
    const confidences: number[] = [];

    for (const col of columns) {
      if (isInternal(col)) continue;

      const values = clusterRows.map((r) => r[col] ?? null);

      const fieldRule: GoldenFieldRule =
        rules.fieldRules[col] ?? { strategy: rules.defaultStrategy as GoldenFieldRule["strategy"] };

      let sources: string[] | undefined;
      let dates: unknown[] | undefined;
      let weights: number[] | undefined;

      if (fieldRule.strategy === "source_priority") {
        sources = clusterRows.map((r) => String(r.__source__ ?? ""));
      }
      if (fieldRule.strategy === "most_recent" && fieldRule.dateColumn) {
        dates = clusterRows.map((r) => r[fieldRule.dateColumn!] ?? null);
      }
      if (qualityScores) {
        weights = rowIds.map(
          (rid) => qualityScores.get(`${rid}:${col}`) ?? 1.0,
        );
      }

      const mergeOpts: MergeFieldOptions = {
        ...(sources !== undefined && { sources }),
        ...(dates !== undefined && { dates }),
        ...(weights !== undefined && { qualityWeights: weights }),
      };
      const result = mergeField(values, fieldRule, mergeOpts);
      confidences.push(result.confidence);

      const sourceRowId =
        result.sourceIndex != null && result.sourceIndex < rowIds.length
          ? rowIds[result.sourceIndex]!
          : rowIds[0]!;

      const candidates = rowIds.map((rid, idx) => {
        const q = qualityScores
          ? (qualityScores.get(`${rid}:${col}`) ?? 1.0)
          : 1.0;
        return { row_id: rid, value: values[idx], quality: q };
      });

      fieldProvenance[col] = {
        value: result.value,
        sourceRowId,
        strategy: fieldRule.strategy,
        confidence: result.confidence,
        candidates,
      };

      goldenRow[col] = result.value;
    }

    const goldenConfidence =
      confidences.length > 0
        ? confidences.reduce((a, b) => a + b, 0) / confidences.length
        : 0.0;

    goldenRow.__golden_confidence__ = goldenConfidence;

    goldenRecords.push(
      goldenRow as Row & {
        __cluster_id__: number;
        __golden_confidence__: number;
      },
    );

    provenanceList.push({
      clusterId: cid,
      clusterQuality: cinfo?.clusterQuality ?? "strong",
      clusterConfidence: cinfo?.confidence ?? 0.0,
      fields: fieldProvenance,
    });
  }

  return { goldenRecords, provenance: provenanceList };
}
